lib/licensed/sources/nuget.rb

```diff
@@ -0,0 +1,226 @@
+# frozen_string_literal: true
+require "json"
+
+module Licensed
+  module Sources
+    # Only supports ProjectReference (project.assets.json) style restore used in .NET Core.
+    # Does not currently support packages.config style restore.
+    class NuGet < Source
+      def self.type
+        "nuget"
+      end
+
+      # Match nuspec license expressions like <license type="expression">MIT</license>
+      class NuspecLicenseExpressionMatcher < Licensee::Matchers::Package
+        LICENSE_EXPRESSION_REGEX = /
+          <license\s*type\s*=\s*\"\s*expression\s*\"\s*>\s*([a-z\-0-9\.]+)\s*<\/license>
+        /ix.freeze
+
+        def license_property
+          return @license_property if defined?(@license_property)
+          match = @file.content.match LICENSE_EXPRESSION_REGEX
+          if match && match[1]
+            @license_property = match[1].downcase
+          end
+        end
+      end
+
+      class NuspecFile < Licensee::ProjectFiles::ProjectFile
+        def possible_matchers
+          [NuspecLicenseExpressionMatcher]
+        end
+      end
+
+      # A pseudo-license file where the license is already known (e.g. inferred from a URL)
+      class KnownLicenseFile < Licensee::ProjectFiles::ProjectFile
+        def initialize(license, metadata)
+          super(license.content, metadata)
+          @license = license
+        end
+
+        def license
+          return @license
+        end
+
+        def confidence
+          100
+        end
+      end
+
+      class NuGetDependency < Licensed::Dependency
+        LICENSE_FILE_REGEX = /<license\s*type\s*=\s*\"\s*file\s*\"\s*>\s*(.*)\s*<\/license>/ix.freeze
+        LICENSE_URL_REGEX = /<licenseUrl>\s*(.*)\s*<\/licenseUrl>/ix.freeze
+
+        def license
+          # By default, multiple of the same/similar licenses (e.g. a package with LICENSE.txt, license, and licenseUrl)
+          # will incorrectly result in an "other" license.
```

Comment:
Yeah, you must be right; one of them must have been classified as `other`.

I think the general issue is that since there's almost always a `licenseUrl` specified, we'll get back a lot of HTML or otherwise low-quality text that's likely to be classified as `other`.

I agree overriding the licensee behavior feels wrong, but it also feels wrong to me to have a significant number of packages that are using the new nuget `license` SPDX expressions get classified as `other` just because they decided to also include a URL to the HTML version of the license.

I'll see how implementing HTML -> text affects this on my sample package set, but I'm guessing it still won't be good enough to avoid `other` classifications.

I'll open an issue on licensee since it seems like `Project.license` should take into account the confidence especially if one is 100 and the other is near-zero.
An alternative if licensee doesn't accept a change here might be to just ignore HTML licenses.

Reply:
At least for my test cases, using ReverseMarkdown seems to work pretty well and I don't think this is a major issue.

Reply:
Actually, I spoke too soon. ReverseMarkdown does a great job at making the text readable, but it still results in lots of 'other' classifications. If I were to always download licenseUrl (regardless of if it or <license> matched), most packages with html licenses then become other.

Reply:
🤔 is this on the test fixture, or a project you're using this with?  if it's on something different, can you put together an example project you can share publicly that I can take a look at?

Reply:
Any package that uses https://licenses.nuget.org is a good example, since it doesn't return the raw license text; it returns HTML with "SPDX identifier" and "License text" sections. e.g. https://licenses.nuget.org/MIT

ReverseMarkdown'ing that page results in fairly readable license text, but is not identified as MIT because of `## SPDX identifier\n\nMIT\n\n## License text\n\n`.

Special-casing licenses.nuget.org is an option, of course. But as I mentioned somewhere - if licenses.nuget.org is used, the package author very likely used `<license>` anyways. Let me see if I can find other examples.

Reply:
http://www.apache.org/licenses/LICENSE-2.0.html is another example. `licensee` recognizes it, but a ReverseMarkdown of the actual page is certainly not the license text:
```
text: "Toggle navigation\n\n- \n\n- [News](/index.html#news)\n- [About&nbsp;](#)\n
    \ - [Overview](/foundation)\n  - [Process](/foundation/how-it-works.html)\n  -
    [Governance](/foundation/governance/)\n  - [The Apache Way](/theapacheway/index.html)\n
    \ - [Membership](/foundation/governance/members.html)\n  - [Community](https://community.apache.org/)\n
...
```
This is a case where the license text is effectively useless and will need to manually be removed, in addition to the incorrect 'other' classification.

Another case is http://www.microsoft.com/web/webpi/eula/net_library_eula_enu.htm. While it's not a recognized license (although I wish it were since it's so common), it would likely have a similar issue since ReverseMarkdown converts links to markdown-style links, leaves some nbsp; around, etc. In this case, ReverseMarkdown produces good enough text to read, but unlikely that it'd be good enough to detect.

Reply:
Per our discussion, I am going to leave the current behavior of downloading the license text only when no other match (package file or license file) is found. After this PR is merged though I'll file an issue to consider revisiting this, either with or without additional changes to `licensee`.